---
title: PGx Analysis Report
pagetitle: PGx Analysis Report
output: 
  rmdformats::readthedown:
    highlight: tango
    toc_depth: 5
    gallery: FALSE
    lightbox: FALSE
    css: style.css

date: "`r format(Sys.time(), '%d %B, %Y')`"
author: Ram Sai Nanduri
urlcolor: blue
linkcolor: blue
---

<style>

body {
  text-align: justify;
  }

h1.title {
  text-align: left;
  }

h4.date {
  text-align: center;
  }


</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  collapse=FALSE
  )
library(dplyr)
library(knitr)
library(kableExtra)
library(ggplot2)
library(rmarkdown)
library(scales)
library(rmdformats)
library(tidyverse)
library(knitr)
source("runReportFunctions.R")
source("Ploty_coverage_plot.R")
```

```{r,echo = FALSE}
tableNo = 1
figureNo = 0
```  

```{r input, include=FALSE, echo=FALSE}

detected_variants <- read.csv(found_variants, sep = "\t", stringsAsFactors = F, row.names = NULL)
missed_variants_df <- read.csv(missed_variants, sep = "\t", stringsAsFactors = F,  row.names = NULL)
haplotype_definitions_df <- read.csv(haplotype_definitions, sep = "\t", stringsAsFactors = F,  row.names = NULL)
clinical_guidelines_df <- read.csv(clinical_guidelines, sep = "\t", stringsAsFactors = F,  row.names = NULL)
interaction_guidelines_df <- read.csv(interaction_guidelines, sep = "\t", stringsAsFactors = F,  row.names = NULL)

```

```{r input2, include=FALSE, echo=FALSE}
# Formating
detected_variants$Zygosity <- sapply(detected_variants$GT, function(x){
   b <- sum(as.numeric(strsplit(x, "[/|]", perl=T)[[1]]))
   return(c("Hetero-", "Homo-")[b])
})
risk_haplotypes <- array()
if (nrow(detected_variants) != 0) {
  detected_variants$Position <- sapply(1:nrow(detected_variants), function(i) paste0(detected_variants[i, 1], ":", detected_variants[i, "POS"]))
  detected_variants <- detected_variants %>% separate("AD", c("Ref.reads", "Alt.reads"), ",", remove=F)
  detected_variants[, "Ref.reads"] <- as.numeric(detected_variants[, "Ref.reads"])
  detected_variants[, "Alt.reads"] <- as.numeric(detected_variants[, "Alt.reads"])
  detected_variants$Haplotype <-
    sapply(detected_variants$ID, function(x) get_haplotypes(x, haplotype_definitions_df))
  columns <- c("Position", "GENE", "ID", "Haplotype", "Zygosity",  "Ref.reads", "Alt.reads", "GQ")
  detected_variants_present <- detected_variants[, columns]
  detected_variants_present[, "Variant Frequency"] <- detected_variants_present[, "Alt.reads"]/(detected_variants_present[, "Alt.reads"] + detected_variants_present[, "Ref.reads"])
  detected_variants_present[, "Possible Duplication"] <- sapply(detected_variants_present[, "Variant Frequency"], risk_duplication)
  faulty_haplotypes <-
          detected_variants_present[detected_variants_present[, "Possible Duplication"] == TRUE, "Haplotype"]
  faulty_haplotypes <- unique(unlist(sapply(faulty_haplotypes, function(x) strsplit(x, "/"))))
} else {
  faulty_haplotypes <- array()
}
```  

# Sample Details
```{r}
header_table <- data.frame(
  row_table = c("Sample ID", "dbSNP version", "Reference genome version"),
  text = c(sample, dbSNP, assembly_genome)
)

kable(header_table, col.names=NULL) %>%
  kable_styling(full_width=T, bootstrap_options="striped") %>%
  column_spec(1, bold=T, border_right=T, width="15em")

```  

# Clinical recommendations based on genotypes
```{r}
clin_columns <- c("gene", "Haplotype1", "Haplotype2", "Guideline")
verbose_columns <- c("Gene", "Haplotype 1", "Haplotype 2", "Clinical Recommendation")
clinical_guidelines_present <- clinical_guidelines_df[, clin_columns]
colnames(clinical_guidelines_present) <- verbose_columns

warning_idx <- which(clinical_guidelines_present[, "Haplotype 1"] %in% faulty_haplotypes | clinical_guidelines_present[, "Haplotype 2"] %in% faulty_haplotypes)

clinical_guidelines_present %>%
    kable(format = "html", escape = F) %>%
    kable_styling() %>%
    row_spec(warning_idx, bold=T, color ="white", background = "red")

```

```{r results="asis"}
if (length(warning_idx) != 0){
  cat("<b><u>One or more variants belonging to haplotype have been flagged as uncertain.\n Do not follow clinical recommendations highlighted in red!</u></b>")
}

```

# Clinical recommendations based on interactions between genotypes
```{r}
if (nrow(interaction_guidelines_df) != 0) {
  interaction_warning_idx <-
          which(sapply(interaction_guidelines_df[,"haplotypes"], function(x) any(strsplit(x, ",") %in% faulty_haplotypes)))
} else {
  interaction_warning_idx <- integer(0)
}

interaction_guidelines_df %>%
    kable(format = "html", escape = F) %>%
    kable_styling() %>%
    row_spec(interaction_warning_idx, bold=T, color ="white", background = "red")
```

# Clinically relevant variants discovered
```{r detected_variants,results='asis',message=FALSE,warning=FALSE}
if (nrow(detected_variants) != 0 ){
  order_columns <- c("GENE", "ID", "Haplotype", "Position", "Zygosity", "Variant Frequency", "Possible Duplication")

  verbose_columns <- c(
    "Gene", "rsID", "Possible Haplotypes", "Position",
    "Zygosity", "Variant Frequency", "Possible Duplication"
  )
  detected_variants_present <- detected_variants_present[, order_columns]
  colnames(detected_variants_present) <- verbose_columns

  detected_variants_present %>%
    kable() %>%
    kable_styling() %>%
    row_spec(which(detected_variants_present[, "Possible Duplication"] == T), bold=T, color="white", background="red")
} else {

cat("<b>There are no variants detected.\n</b>")
  
}
```

# Read depth at wild-type clinical targets
```{r coverage_wildtype}
columns <- c("ID", "Haplotype", "Locus", "Total_Depth")
missed_variants_df$Haplotype <-
  sapply(missed_variants_df$ID, function(x) get_haplotypes(x, haplotype_definitions_df))
missed_variants_df <- missed_variants_df[, columns]
colnames(missed_variants_df) <- c("rsID", "Haplotype", "Position", "Read Depth")
missed_variants_df %>%
  kable() %>%
  kable_styling() %>%
  row_spec(which(missed_variants_df[, "Read Depth"] < 100), bold=T, color ="white", background = "red") # hardcoded bad
```


# Read depth at targets
```{r}
bait_table_file <- target_regions_bed
genes <- read.table(bait_table_file, sep = "\t", as.is=T, check.names = F,
                           col.names = c("Chr", "Start", "End", "Name")) %>%
    separate(Name, c("Target", "Gene"), "_") %>% {unique(.$Gene)}
plt  <- htmltools::tagList()
i    <- 0
for (gene in genes){
  i <- i + 1
  exon_table_file <- paste0(data_location, "/exons/", gene,".csv")
  gdf_file <- depth_file

  plt[[i]] <- as_widget(get_ploty_depth_plot(
    found_variants,
    target_rsid,
    bait_table_file,
    gdf_file,
    exon_table_file,
    gene))
}
```

```{r}
plt
```

# Appendix
Here you will find general information about genes and their connection to drugs, as well as the type of library preparation, sequencing instruments and the like.

